flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBase.java

```diff
@@ -140,7 +152,12 @@ public void invoke(IN value) throws Exception {
             semaphore.release();
             throw e;
         }
-        Futures.addCallback(result, callback);
+        if (synchronousWrites) {
+            result.get(timeout, TimeUnit.SECONDS);
+            semaphore.release();
```

Comment:
If this times out the semaphore is never released.

Reply:
thanks `semaphore.release();` should be in a finally block